File: windows/SeedSeeker/ResultNavigation.cs
namespace SeedSeeker;

/// <summary>
/// Navigation through the ordered list of search-result seeds while scouting.
///
/// The seed detail pane can be reached either from a search result or by
/// typing a seed by hand; navigation is only meaningful in the first case, so
/// every helper returns null when the current seed is not a search result.
/// Pure so the logic stays testable off-Windows (see SeedSeeker.Tests).
/// </summary>
public static class ResultNavigation
{
    /// <summary>0-based index of <paramref name="seed"/> within <paramref name="seeds"/>, or null when it is not one of them.</summary>
    public static int? IndexOf(IReadOnlyList<string> seeds, string? seed)
    {
        if (string.IsNullOrEmpty(seed)) return null;
        for (var index = 0; index < seeds.Count; index++)
            if (seeds[index] == seed) return index;
        return null;
    }

    /// <summary>
    /// Index of the result <paramref name="delta"/> steps away from
    /// <paramref name="seed"/>, clamped to the list ends. Null when the seed
    /// is not a search result or the step would not move (already at the
    /// first or last result).
    /// </summary>
    public static int? Step(IReadOnlyList<string> seeds, string? seed, int delta)
    {
        if (IndexOf(seeds, seed) is not int index) return null;
        var target = Math.Clamp(index + delta, 0, seeds.Count - 1);
        return target == index ? null : target;
    }
}

File: macos/SeedSeeker/Sources/SeedSeekerKit/ResultNavigation.swift
import Foundation

/// Navigation through the ordered list of search-result seeds while scouting.
///
/// The seed detail pane can be reached either from a search result or by
/// typing a seed by hand; navigation is only meaningful in the first case, so
/// every helper returns nil when the current seed is not a search result.
public enum ResultNavigation {
    /// 0-based index of `seed` within `seeds`, or nil when it is not one of them.
    public static func position(of seed: String?, in seeds: [String]) -> Int? {
        guard let seed, !seed.isEmpty else { return nil }
        return seeds.firstIndex(of: seed)
    }

    /// The seed `offset` steps away from `seed` in the results, clamped to the
    /// list ends. Nil when `seed` is not a search result or the step would not
    /// move (already at the first or last result).
    public static func seed(from seed: String?, in seeds: [String], offset: Int) -> String? {
        guard let index = position(of: seed, in: seeds) else { return nil }
        let target = min(max(index + offset, 0), seeds.count - 1)
        return target == index ? nil : seeds[target]
    }
}
